Spurious ClassCastException in HashMap used in CleanerProvider

### Current Behavior

I got a strange error with a ClassCastException when adding something to a map in the CleanerProvider code:

```
Pipe [Extract From ZIP File] caught exception: Exception on transforming input: (ClassCastException) class java.util.HashMap$Node cannot be cast to class java.util.HashMap$TreeNode (java.util.HashMap$Node and java.util.HashMap$TreeNode are in module java.base of loader 'bootstrap')
	at org.frankframework.senders.IbisJavaSender.sendMessage(IbisJavaSender.java:137)
	at org.frankframework.larva.SenderThread.run(SenderThread.java:76)
Caused by: nl.nn.adapterframework.dispatcher.RequestProcessorException: RequestProcessor [ibis4test-JmsBytesSender] caught exception
	at nl.nn.adapterframework.dispatcher.DispatcherManagerImpl.processRequest(DispatcherManagerImpl.java:128)
	at org.frankframework.senders.IbisJavaSender.sendMessage(IbisJavaSender.java:133)
	... 1 more

Caused by: java.lang.ClassCastException: class java.util.HashMap$Node cannot be cast to class java.util.HashMap$TreeNode (java.util.HashMap$Node and java.util.HashMap$TreeNode are in module java.base of loader 'bootstrap')
	at java.base/java.util.HashMap$TreeNode.moveRootToFront(HashMap.java:1986)
	at java.base/java.util.HashMap$TreeNode.treeify(HashMap.java:2102)
	at java.base/java.util.HashMap.treeifyBin(HashMap.java:770)
	at java.base/java.util.HashMap.putVal(HashMap.java:642)
	at java.base/java.util.HashMap.put(HashMap.java:610)
	at org.frankframework.util.CleanerProvider.register(CleanerProvider.java:84)
	at org.frankframework.core.PipeLineSession.createCloseAction(PipeLineSession.java:104)
	at org.frankframework.core.PipeLineSession.<init>(PipeLineSession.java:99)
	at org.frankframework.pipes.IteratingPipe$ItemCallback.handleItem(IteratingPipe.java:278)
	at org.frankframework.pipes.IteratingPipe.iterateOverInput(IteratingPipe.java:188)
	at org.frankframework.pipes.IteratingPipe.sendMessage(IteratingPipe.java:477)
	... 105 more

```

It appears that the LeakLog map is misbehaving. I can only explain this as a threading issue as the map is not threadsafe but could be updated from multiple threads. Oops.

### Expected Behavior

No exceptions when adding to leaklog.

### Environment Information

Version 9.3-snapshot.

Can likely also occur in 9.0 and 9.2.

### Steps To Reproduce

_No response_

### Configuration

```xml

```

### Input

_No response_

### What database are you using?

_No response_

### What browsers are you seeing the problem on?

_No response_

### Relevant Log Output

```console

```

### Anything else?

Unfortunately it will be very hard to verify a fix.

To reduce performance impact of using a ConcurrentHashMap we will no longer use the LeakLog for each PipeLineSession (and in backports to older versions, for each Message).

--- a/core/src/main/java/org/frankframework/core/PipeLineSession.java
+++ b/core/src/main/java/org/frankframework/core/PipeLineSession.java
@@ -16,6 +16,7 @@
 package org.frankframework.core;
 
 import java.io.IOException;
+import java.lang.ref.Cleaner;
 import java.time.Instant;
 import java.time.temporal.Temporal;
 import java.util.Collections;
@@ -79,6 +80,7 @@ public class PipeLineSession extends HashMap<String,Object> implements AutoClose
 	private ISecurityHandler securityHandler = null;
 
 	private transient PipeLineSessionCloseAction closeAction;
+	private transient Cleaner.Cleanable cleanable;
 
 	// closeables.keySet is a List of wrapped resources. The wrapper is used to unschedule them, once they are closed by a regular step in the process.
 	// Values are labels to help debugging
@@ -101,7 +103,7 @@ public class PipeLineSession extends HashMap<String,Object> implements AutoClose
 
 	private void createCloseAction() {
 		closeAction = new PipeLineSessionCloseAction(this.closeables);
-		CleanerProvider.register(this, closeAction);
+		cleanable = CleanerProvider.CLEANER.register(this, closeAction);
 	}
 
 	public void setExitState(PipeLine.ExitState state, Integer code) {
@@ -483,7 +485,7 @@ public class PipeLineSession extends HashMap<String,Object> implements AutoClose
 	@Override
 	public void close() {
 		LOG.debug("Closing PipeLineSession");
-		CleanerProvider.clean(closeAction);
+		cleanable.clean();
 	}
 
 	private static class PipeLineSessionCloseAction implements Runnable {

--- a/core/src/main/java/org/frankframework/core/PipeLineSession.java
+++ b/core/src/main/java/org/frankframework/core/PipeLineSession.java
@@ -77,9 +77,7 @@ public class PipeLineSession extends HashMap<String,Object> implements AutoClose
 	public static final String EXIT_STATE_CONTEXT_KEY="exitState";
 	public static final String EXIT_CODE_CONTEXT_KEY="exitCode";
 
-	private ISecurityHandler securityHandler = null;
-
-	private transient PipeLineSessionCloseAction closeAction;
+	private transient ISecurityHandler securityHandler = null;
 	private transient Cleaner.Cleanable cleanable;
 
 	// closeables.keySet is a List of wrapped resources. The wrapper is used to unschedule them, once they are closed by a regular step in the process.
@@ -102,6 +100,7 @@ public class PipeLineSession extends HashMap<String,Object> implements AutoClose
 	}
 
 	private void createCloseAction() {
+  PipeLineSessionCloseAction closeAction;
 		closeAction = new PipeLineSessionCloseAction(this.closeables);
 		cleanable = CleanerProvider.CLEANER.register(this, closeAction);
 	}

--- a/core/src/main/java/org/frankframework/stream/OverflowToDiskOutputStream.java
+++ b/core/src/main/java/org/frankframework/stream/OverflowToDiskOutputStream.java
@@ -50,6 +50,7 @@ public class OverflowToDiskOutputStream extends OutputStream implements AutoClos
 	private boolean closed = false;
 
 	private CleanupFileAction cleanupFileAction;
+	private Cleaner.Cleanable cleanable;
 
 	/**
 	 * The number of bytes in the buffer. This value is always in the range
@@ -89,7 +90,7 @@ public class OverflowToDiskOutputStream extends OutputStream implements AutoClos
 	 */
 	private void createCleanerAction(final Path path, final Closeable closable) {
 		cleanupFileAction = new CleanupFileAction(path, closable);
-		CleanerProvider.register(this, cleanupFileAction);
+		cleanable = CleanerProvider.register(this, cleanupFileAction);
 	}
 
 	private static class CleanupFileAction implements Runnable {
@@ -204,7 +205,7 @@ public class OverflowToDiskOutputStream extends OutputStream implements AutoClos
 			//Since we were successfully able to create a PathMessage (which will cleanup the file on close) remove the reference here.
 			fileLocation = null;
 			cleanupFileAction.shouldClean = false;
-			CleanerProvider.clean(cleanupFileAction);
+			CleanerProvider.clean(cleanable);
 
 			return result;
 		} else {

--- a/core/src/main/java/org/frankframework/stream/SerializableFileReference.java
+++ b/core/src/main/java/org/frankframework/stream/SerializableFileReference.java
@@ -27,6 +27,7 @@ import java.io.Reader;
 import java.io.Serial;
 import java.io.Serializable;
 import java.io.StringReader;
+import java.lang.ref.Cleaner;
 import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -59,7 +60,7 @@ public class SerializableFileReference implements Serializable, AutoCloseable {
 	private String charset;
 	@Getter private transient Path path;
 	private transient CleanupFileAction cleanupFileAction;
-
+	private transient Cleaner.Cleanable cleanable;
 
 	/**
 	 * Create a new {@link SerializableFileReference} from the given {@link InputStream}. The {@link InputStream} will be copied
@@ -148,7 +149,7 @@ public class SerializableFileReference implements Serializable, AutoCloseable {
 
 	private void createCleanerAction(final Path path) {
 		cleanupFileAction = new CleanupFileAction(path);
-		CleanerProvider.register(this, cleanupFileAction);
+		cleanable = CleanerProvider.register(this, cleanupFileAction);
 	}
 
 	private static class CleanupFileAction implements Runnable {
@@ -202,8 +203,8 @@ public class SerializableFileReference implements Serializable, AutoCloseable {
 
 	@Override
 	public void close() {
-		if (cleanupFileAction != null) {
-			CleanerProvider.clean(cleanupFileAction);
+		if (cleanable != null) {
+			CleanerProvider.clean(cleanable);
 		}
 	}
 

--- a/core/src/main/java/org/frankframework/stream/SerializableFileReference.java
+++ b/core/src/main/java/org/frankframework/stream/SerializableFileReference.java
@@ -53,13 +53,13 @@ public class SerializableFileReference implements Serializable, AutoCloseable {
 	public static final String TEMP_MESSAGE_DIRECTORY = "temp-messages";
 
 	@Serial private static final long serialVersionUID = 1L;
+	@SuppressWarnings("java:S115")
 	private static final long customSerializationVersion = 1L;
 
 	private long size = Message.MESSAGE_SIZE_UNKNOWN;
 	@Getter private boolean binary;
 	private String charset;
 	@Getter private transient Path path;
-	private transient CleanupFileAction cleanupFileAction;
 	private transient Cleaner.Cleanable cleanable;
 
 	/**
@@ -148,7 +148,7 @@ public class SerializableFileReference implements Serializable, AutoCloseable {
 	}
 
 	private void createCleanerAction(final Path path) {
-		cleanupFileAction = new CleanupFileAction(path);
+		CleanupFileAction cleanupFileAction = new CleanupFileAction(path);
 		cleanable = CleanerProvider.register(this, cleanupFileAction);
 	}
 

--- a/core/src/main/java/org/frankframework/util/CleanerProvider.java
+++ b/core/src/main/java/org/frankframework/util/CleanerProvider.java
@@ -39,7 +39,8 @@ public class CleanerProvider {
 
 	private static final ConcurrentMap<Integer, CleaningActionWrapper> CLEANER_MAP = new ConcurrentHashMap<>();
 	private static final Map<LeakedResourceException, AtomicInteger> LEAK_MAP = new HashMap<>(); // Should be used only in Cleaner thread so doesn't need to be thread-safe
-	private static final Cleaner CLEANER = Cleaner.create();
+
+	public static final Cleaner CLEANER = Cleaner.create();
 
 	static {
 		// Shutdown hook to generate report of leaks at end of process

--- a/core/src/main/java/org/frankframework/util/CleanerProvider.java
+++ b/core/src/main/java/org/frankframework/util/CleanerProvider.java
@@ -21,6 +21,8 @@ import java.util.Arrays;
 import java.util.Comparator;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import org.apache.logging.log4j.LogManager;
@@ -35,8 +37,8 @@ import lombok.NoArgsConstructor;
 public class CleanerProvider {
 	private static final Logger LEAK_LOG = LogManager.getLogger("LEAK_LOG");
 
-	private static final Map<Integer, CleaningActionWrapper> CLEANER_MAP = new HashMap<>();
-	private static final Map<LeakedResourceException, AtomicInteger> LEAK_MAP = new HashMap<>();
+	private static final ConcurrentMap<Integer, CleaningActionWrapper> CLEANER_MAP = new ConcurrentHashMap<>();
+	private static final Map<LeakedResourceException, AtomicInteger> LEAK_MAP = new HashMap<>(); // Should be used only in Cleaner thread so doesn't need to be thread-safe
 	private static final Cleaner CLEANER = Cleaner.create();
 
 	static {

--- a/core/src/main/java/org/frankframework/util/CleanerProvider.java
+++ b/core/src/main/java/org/frankframework/util/CleanerProvider.java
@@ -82,11 +82,13 @@ public class CleanerProvider {
 	 * @param cleaningAction Cleaning Action to register
 	 * @return ID of the cleaning action, which can be used to execute it.
 	 */
-	public static int register(Object obj, Runnable cleaningAction) {
+	public static Cleaner.Cleanable register(Object obj, Runnable cleaningAction) {
 		CleaningActionWrapper wrapper = new CleaningActionWrapper(obj, cleaningAction);
-		CLEANER_MAP.put(wrapper.actionId, wrapper);
-		wrapper.cleanable = CLEANER.register(obj, wrapper);
-		return wrapper.actionId;
+		Cleaner.Cleanable cleanable = CLEANER.register(obj, wrapper);
+		int actionId = getActionId(cleanable);
+		wrapper.actionId = actionId;
+		CLEANER_MAP.put(actionId, wrapper);
+		return cleanable;
 	}
 
 	/**
@@ -94,47 +96,30 @@ public class CleanerProvider {
 	 * is out of scope, for instance from a close-method.
 	 * This de-registers the cleaning action and marks it so that it is not counted as leaked.
 	 *
-	 * @param cleaningAction Cleaning Action to execute.
+	 * @param cleanable Cleaning Action to execute.
 	 */
-	public static void clean(Runnable cleaningAction) {
-		if (cleaningAction == null) {
+	public static void clean(Cleaner.Cleanable cleanable) {
+		if (cleanable == null) {
 			return;
 		}
-		int actionId = getActionId(cleaningAction);
+		int actionId = getActionId(cleanable);
 		CleaningActionWrapper wrapper = CLEANER_MAP.get(actionId);
 		if (wrapper != null) {
 			wrapper.cleaned = true;
-			wrapper.cleanable.clean();
-		} else {
-			cleaningAction.run();
 		}
-	}
-
-	/**
-	 * Execute cleaning action by ID.
-	 *
-	 * @param actionId ID of the cleaning action.
-	 */
-	public static void clean(int actionId) {
-		CleaningActionWrapper wrapper = CLEANER_MAP.get(actionId);
-		if (wrapper != null) {
-			wrapper.cleaned = true;
-			wrapper.cleanable.clean();
-		}
+		cleanable.clean();
 	}
 
 	private static class CleaningActionWrapper implements Runnable {
 		private final Runnable cleaningAction;
-		private final int actionId;
 		private final String owningClassName;
 		private final boolean isProxyClass;
 		private final LeakedResourceException creationTrace;
+		private int actionId;
 		private boolean cleaned = false;
-		private Cleaner.Cleanable cleanable;
 
 		private CleaningActionWrapper(Object owner, Runnable cleaningAction) {
 			this.cleaningAction = cleaningAction;
-			actionId = getActionId(cleaningAction);
 			owningClassName = owner.getClass().getName();
 			isProxyClass = Proxy.isProxyClass(owner.getClass()) || owningClassName.contains(org.springframework.util.ClassUtils.CGLIB_CLASS_SEPARATOR);
 			creationTrace = new LeakedResourceException(owner);
@@ -156,8 +141,8 @@ public class CleanerProvider {
 		}
 	}
 
-	private static int getActionId(Runnable cleaningAction) {
-		return System.identityHashCode(cleaningAction);
+	private static int getActionId(Cleaner.Cleanable cleanable) {
+		return System.identityHashCode(cleanable);
 	}
 
 	private static class LeakedResourceException extends RuntimeException {

--- a/core/src/main/java/org/frankframework/util/CleanerProvider.java
+++ b/core/src/main/java/org/frankframework/util/CleanerProvider.java
@@ -1,5 +1,5 @@
 /*
-   Copyright 2024 WeAreFrank!
+   Copyright 2024-2025 WeAreFrank!
 
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.

--- a/core/src/main/java/org/frankframework/util/flow/graphviz/GraphvizEngine.java
+++ b/core/src/main/java/org/frankframework/util/flow/graphviz/GraphvizEngine.java
@@ -1,5 +1,5 @@
 /*
-   Copyright 2018-2020, 2022 WeAreFrank!
+   Copyright 2018-2020, 2022-2025 WeAreFrank!
 
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
@@ -16,6 +16,7 @@
 package org.frankframework.util.flow.graphviz;
 
 import java.io.IOException;
+import java.lang.ref.Cleaner;
 import java.net.URL;
 import java.util.Arrays;
 
@@ -48,7 +49,7 @@ public class GraphvizEngine {
 	// Available JS Engines. Lower index has priority.
 	private static final String[] engines = AppConstants.getInstance().getString("flow.javascript.engines", "org.frankframework.javascript.J2V8").split(",");
 
-	private final int cleaningActionId;
+	private final Cleaner.Cleanable cleanable;
 	private Engine engine;
 	private String graphvizVersion = AppConstants.getInstance().getProperty("graphviz.js.version", "2.0.0");
 
@@ -84,7 +85,7 @@ public class GraphvizEngine {
 		getEngine();
 
 		CleanupEngineAction cleanupEngineAction = new CleanupEngineAction(engine);
-		cleaningActionId = CleanerProvider.register(this, cleanupEngineAction);
+		cleanable = CleanerProvider.CLEANER.register(this, cleanupEngineAction);
 	}
 
 	private static class CleanupEngineAction implements Runnable {
@@ -176,7 +177,7 @@ public class GraphvizEngine {
 	 * {@link GraphvizEngine GraphvisEngines}. This method ensures that the used Javascript engine is destroyed properly.
 	 */
 	public void close() {
-		CleanerProvider.clean(cleaningActionId);
+		cleanable.clean();
 	}
 
 	private String getVisJsWrapper() {

